crates/red_knot_python_semantic/src/types/class.rs

```diff
@@ -391,16 +393,40 @@ impl<'db> Class<'db> {
                     );
                 }
                 ClassBase::Class(class) => {
-                    if let member @ SymbolAndQualifiers(Symbol::Type(_, _), _) =
+                    if let member @ SymbolAndQualifiers(Symbol::Type(ty, boundness), _) =
                         class.own_instance_member(db, name)
                     {
-                        return member;
+                        if boundness == Boundness::Bound {
+                            if union.is_empty() {
+                                // Short-circuit, no need to allocate inside the union builder
+                                return member;
+                            }
```

Comment:
is it worth generalising this optimisation inside the `UnionBuilder` rather than special-casing single-element unions here? The `UnionBuilder::elements` field is currently just `Vec<Type>`, but if it was something like `SmallVec<[Type; 1]>`, that would avoid allocating inside the union builder if there was only 1 element

Reply:
> Although using `from_elements` might be challenging here.

yeah, I think it's probably not practical to use it here unfortunately